### ISSUE TITLE
Negative cursor position crash

### ISSUE THREAD

**Issue by electrodude** [vim/vim#674]
Typing `:set ruler<CR>asdf<Esc>^<C-X>10hasdfasdfasdfasdf...` into Vim will crash it by inserting text at a negative cursor position.

First, do `:set ruler<CR>` to show the cursor position and type some text `asdf<Esc>`.  Then, do `g0<C-X>10h` and notice that the cursor X has gone negative `1,-10-1`.  Attempting to insert text `asdfasdfasdf...` with the cursor negative then crashes Vim:

```
Vim: Caught deadly signal ABRT
Vim: preserving files...
Vim: Finished.
Aborted
```

This also works with `<C-A>` (increment number under cursor) instead of `<C-X>` (decrement number under cursor).

Basically, the bug is that if <C-X> doesn't have a number to decrement, it moves the cursor left, even if it's already at column 1, in which case the cursor goes to 0 and then negative if you push `h`.

Also, neither `A` nor `l` with a negative cursor moves it right again, but `I` and `$` do.

**Comment by chrisbra** [comment]
Does this happen with vim -u NONE -N I tried, but it aborted correctly after <c-x>
What version did you use and did you use vim -u NONE -N

**Comment by electrodude** [comment]
Yes, this happens with `vim -u NONE -N`.

I'm running Gentoo:

```
=app-editors/vim-7.4.769:0 X acl cscope gpm lua luajit nls perl python tcl vim-pager -debug -minimal -racket -ruby -selinux
```

Version info:

```
$ vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Feb 29 2016 22:24:40)
Included patches: 1-769
Modified by Gentoo-7.4.769
Compiled by [REDACTED - me]
Huge version without GUI.  Features included (+) or not (-):
+acl             +farsi           +mouse_netterm   +syntax
+arabic          +file_in_path    +mouse_sgr       +tag_binary
+autocmd         +find_in_path    -mouse_sysmouse  +tag_old_static
-balloon_eval    +float           +mouse_urxvt     -tag_any_white
-browse          +folding         +mouse_xterm     +tcl
++builtin_terms  -footer          +multi_byte      +terminfo
+byte_offset     +fork()          +multi_lang      +termresponse
+cindent         +gettext         -mzscheme        +textobjects
+clientserver    -hangul_input    +netbeans_intg   +title
+clipboard       +iconv           +path_extra      -toolbar
+cmdline_compl   +insert_expand   +perl            +user_commands
+cmdline_hist    +jumplist        +persistent_undo +vertsplit
+cmdline_info    +keymap          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python/dyn      +viminfo
+cscope          +lispindent      +python3/dyn     +vreplace
+cursorbind      +listcmds        +quickfix        +wildignore
+cursorshape     +localmap        +reltime         +wildmenu
+dialog_con      +lua             +rightleft       +windows
+diff            +menu            -ruby            +writebackup
+digraphs        +mksession       +scrollbind      +X11
-dnd             +modify_fname    +signs           +xfontset
-ebcdic          +mouse           +smartindent     -xim
+emacs_tags      -mouseshape      -sniff           +xsmp_interact
+eval            +mouse_dec       +startuptime     +xterm_clipboard
+ex_extra        +mouse_gpm       +statusline      -xterm_save
+extra_search    -mouse_jsbterm   -sun_workshop    -xpm
   system vimrc file: "/etc/vim/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  fall-back for $VIM: "/usr/share/vim"
Compilation: x86_64-pc-linux-gnu-gcc -c -I. -Iproto -DHAVE_CONFIG_H     -O2 -pipe -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1      -I/usr/include  -D_REENTRANT=1  -D_THREAD_SAFE=1  -D_LARGEFILE64_SOURCE=1  
Linking: x86_64-pc-linux-gnu-gcc   -Wl,-E  -Wl,-O1 -L/usr/local/lib -Wl,--as-needed -o vim    -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -lncurses -lelf -lnsl    -lacl -lattr -lgpm -ldl  -L/usr/lib/x86_64-linux-gnu -lluajit-5.1 -Wl,-E -Wl,-O1 -Wl,--as-needed  -L/usr/lib64/perl5/5.20.2/x86_64-linux/CORE -lperl -lnsl -ldl -lm -lcrypt -lutil -lc   -ltcl8.5 -ldl -lpthread -lieee -lm
```

**Comment by chrisbra** [comment]
Can you please try with a more recent Vim? There have been some patches regarding incrementing numbers be done and I suspect this problem is already fixed. Thanks.

**Comment by electrodude** [comment]
I tried Vim 7.4.1342, the latest version Gentoo has, and the bug is gone.  Thanks.
